airbyte-workers/src/main/java/io/airbyte/workers/storage/StateStore.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright (c) 2021 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.workers.storage;
+
+import io.airbyte.commons.json.Jsons;
+import io.airbyte.config.State;
+import io.airbyte.config.storage.CloudStorageConfigs.GcsConfig;
+import io.airbyte.config.storage.CloudStorageConfigs.MinioConfig;
+import io.airbyte.config.storage.CloudStorageConfigs.S3Config;
+import io.airbyte.config.storage.DefaultGcsClientFactory;
+import io.airbyte.config.storage.DefaultS3ClientFactory;
+import io.airbyte.config.storage.MinioS3ClientFactory;
+import java.nio.file.Path;
+import java.util.Optional;
+import java.util.UUID;
+
+/**
+ * Leverages the CloudDocumentStore. The root directory is state and then each activityRunId is the
+ * key.
+ */
+public class StateStore {
+
+  private static final Path STATE_ROOT = Path.of("state");
+
+  private final CloudDocumentStoreClient documentStoreClient;
+
+  public StateStore s3(final S3Config config) {
+    return new StateStore(new S3CloudDocumentStoreClient(
+        new DefaultS3ClientFactory(config).get(),
+        config.getBucketName(),
+        STATE_ROOT));
+  }
+
+  public StateStore minio(final MinioConfig config) {
+    return new StateStore(new S3CloudDocumentStoreClient(
+        new MinioS3ClientFactory(config).get(),
+        config.getBucketName(),
+        STATE_ROOT));
+  }
+
+  public StateStore gcs(final GcsConfig config) {
+    return new StateStore(new GcsCloudDocumentStoreClient(
+        new DefaultGcsClientFactory(config).get(),
+        config.getBucketName(),
+        STATE_ROOT));
+  }
+
+  public StateStore dockerCompose(final Path workspaceMount) {
+    return new StateStore(new DockerComposeDocumentStoreClient(workspaceMount));
+  }
+
+  public StateStore(final CloudDocumentStoreClient documentStoreClient) {
+    this.documentStoreClient = documentStoreClient;
+  }
+
+  /**
+   * Set the state for an activity run. Overwrites existing state if present.
+   *
+   * @param activityRunId - id to associate state with
+   * @param state - state to persist
+   */
+  void setState(final UUID activityRunId, final State state) {
```

Comment:
We can change this in a different PR but `ReplicationOutput` is likely the first interface not just `State`.

Reply:
discussed offline. will just call this a `WorkerStore` and it will use json. then the scheduler project can build on top